### ISSUE TITLE
Add `{deserialize,serialize}_{keys,values}_with` support to serde-derive

### ISSUE THREAD

**Issue by sfackler** [serde-rs/serde#723]
These would be equivalent to Jackson's `@Serialize(keysUsing=...)` etc. Now that we have stateful deserialization, this can be implemented in a pretty straightforward way.

I believe we'd want to support map keys and values as well was seq and option values.

**Comment by dtolnay** [comment]
Ideally I would like to find an approach that composes better than keysUsing.

```rust
#[derive(Deserialize)]
struct S {
    #[serde(deserialize_with = "my_key")]
    key: K,

    #[serde(deserialize_with = "my_value")]
    value: V,

    #[serde(???)]
    opt_map: Option<Map<K, V>>,
}
```

**Comment by dtolnay** [comment]
#576 has an approach based on a helper for generating ordinary deserialize_with functions, rather than using a slate of new attributes.

**Comment by dtolnay** [comment]
This could be neat:

```rust
#[derive(Deserialize)]
struct S {
    #[serde(deserialize_with = "my_key")]
    key: K,

    #[serde(deserialize_with = "my_value")]
    value: V,

    #[serde(deserialize_with = "my_opt_map")]
    opt_map: Option<Map<K, V>>,
}

fn my_map<'de, D>(deserializer: D) -> Result<Map<K, V>, D::Error>
    where D: Deserializer<'de>
{
    deserialize_map_with!(my_key, my_value)(deserializer)
}

fn my_opt_map<'de, D>(deserializer: D) -> Result<Option<Map<K, V>>, D::Error>
    where D: Deserializer<'de>
{
    deserialize_option_with!(my_map)(deserializer)
}
```

**Comment by dtolnay** [comment]
Another possible composable approach:

```rust
#[derive(Deserialize)]
struct S {
    #[serde(deserialize_with = "my_k")]
    k: K,

    #[serde(deserialize_with = "option!(my_k)")]
    opt: Option<K>,

    #[serde(deserialize_with = "option!(map!(my_k, my_v))")]
    opt_map: Option<Map<K, V>>,

    #[serde(deserialize_with = "map!(_, my_v)")]
    map: Map<u8, V>,
}
```

**Comment by dtolnay** [comment]
This needs to support all the wrapper types too: Rc, Arc, Cell, RefCell, Mutex, RwLock.

**Comment by daboross** [comment]
Would a syntax like this also want to support custom key/values within a custom map deserializer?

Like

```
#[derive(Deserialize)]
struct S {
    #[serde(deserialize_with = "my_k")]
    k: K,

    #[serde(deserialize_with = "option!(my_k)")]
    opt: Option<K>,

    #[serde(deserialize_with = "my_map")]
    map: Map<u8, u8>,

    #[serde(deserialize_with = "option!(map!(my_k, my_v))")]
    opt_map: Option<Map<K, V>>,

    #[serde(deserialize_with = "map!(_, my_v)")]
    map: Map<u8, V>,

    #[serde(deserialize_with = "map::<my_map>!(my_k, my_v)")]
}
```
or would this not be feasible at all with the current Deserializer architecture?

**Comment by dtolnay** [comment]
Yes, probably by means of a trait that is implemented for all types that support `map!`, a different one that is implemented for types that support `option!`, etc.

**Comment by daboross** [comment]
hm, ok. Would that mean the `my_map` deserializer would have to return a type which implements some trait then?

I mean `deserialize_with` is often used with types from other libraries, where implementing a trait on the type isn't possible - just trying to hash out the problem here.

For a syntax alternative, what would you think of something like "inner" attribute, like this?
```rust
#[derive(Deserialize)]
struct S {
    #[serde(deserialize_with = "my_k")]
    k: K,

    #[serde(inner(K, deserialize_with = "my_k"))]
    opt: Option<K>,

    #[serde(deserialize_with = "my_map")]
    map: Map<u8, u8>,

    #[serde(inner(K, deserialize_with = "my_k"))]
    #[serde(inner(V, deserialize_with = "my_v"))]
    opt_map: Option<Map<K, V>>,

    #[serde(inner(V, deserialize_with = "my_v"))]
    map: Map<u8, V>,

    #[serde(deserialize_with = "my_map")]
    #[serde(inner(K, deserialize_with = "my_k"))]
    #[serde(inner(V, deserialize_with = "my_v"))]
    map: Map<K, V>,
}
```

If this kind of syntax would be allowed in attributes, and if it relatively matched what we can feasibly make happen, it would also provide an intuitive way to include #914:

```rust
#[derive(Deserialize)]
struct S {
    #[serde(inner(Cow<'a, str>, borrow))]
    cow: Vec<Cow<'a, str>>,
}
```

**Comment by daboross** [comment]
Sorry, I think I was just on the complete wrong track there. My apologies for not researching how all of this actually works and thinking about it before commenting!

I think I can agree now that a trait is probably the best way to do this.

**Comment by HighCommander4** [comment]
I'm pretty new to Rust and Serde, but since you asked (in e.g. #999 and #1005) for people's thoughts on a design, here's how I would _like_ this to work as a user:

```
mod remote {   // the remote crate
  struct Foo {    // the remote struct I'm using
    ...
  }
}
////////////////
mod my {   // my crate
  #[derive(Serialize, Deserialize)]
  struct FooDef {  // a struct with identical fields to Foo
    ...
  }

  // A macro invocation that emits some declarations that make
  // it so that, within this module, _every_ use of Foo is serialized
  // as if it had been annotated with #[serde(with = "FooDef")]
  serde_serialize_as!(Foo, FooDef);

  #[derive(Serialize, Deserialize)]
  struct MyStruct {
    field1: Foo,   // works, no additional annotation necessary
    field2: Option<Foo>,   // also works
    field3: HashMap<Foo, String>   // also works
  }
}
```

I don't know enough about Rust macros and Serde to say whether this is actually implementable.

**Comment by Keats** [comment]
I guess no design has been decided yet? I have a struct that has a `Option<toml::value::Datetime>` but I'd like to store it as a `String` instead since I don't want to care about TOML once it's loaded.

```rust
fn from_toml_datetime<'de, D>(deserializer: D) -> StdResult<Option<String>, D::Error>
    where
        D: Deserializer<'de>,
{
    toml::value::Datetime::deserialize(deserializer)
        .map(|s| Some(s.to_string()))
}
```
was my attempt at `#[serde(deserialize_with = "from_toml_datetime")]`

The above was my intuition so I would vote for something like that, if it's doable.

Is there a way to do that at all currently?

**Comment by daboross** [comment]
@Keats that looks like it should work correctly. Is a field marked with `#[serde(default, deserialize_with = "from_toml_datetime")]` not working?

**Comment by daboross** [comment]
What does anyone think of including non-(de) serialize_with attributes in this issue?

I'm currently running into a situation where it would make sense to use a `HashMap<i64, Cow<'a, [u8]>>` with all inner `Cow`'s borrowed. However, `#[serde(borrow)]` only affects the outermost type, just like (de) `serialize_with`.

**Comment by zbraniecki** [comment]
Is there any way to workaround it? I have a struct like:

```rust
struct Snippet {
    annotations: Vec<Annotation>
}

#[derive(Deserialize)]
#[serde(remote = "Snippet")]
struct SnippetDef {
  #[serde(with = "AnnotationDef")]
  annotations: Vec<Annotations>
}
```

and that of course doesn't work. How should I solve it until this is fixed?

**Comment by daboross** [comment]
@zbraniecki 

My understanding is that you'd work around it by making a method, roughly:

```rust
fn deserialize_annotation_vec<'de, D>(deserializer: D) -> Result<Vec<Annotation>, D::Error> {
    struct AnnotationVecVisitor;
    impl<'de> Visitor<'de> for AnnotationVecVisitor {
        type Value = Vec<Annotation>;
        fn expecting(&self, f: &mut fmt::Formatter) -> fmt::Result { write!(f, "a list of annotations") }
        fn visit_seq<A: SeqAccess<'de>>(self, seq: A) -> Result<Vec<Annotation>, A::Error> {
            let mut vec = Vec::with_capacity(cmp::min(seq.size_hint().unwrap_or(0), 4096));
            while let Some(v) = seq.next_element()? {
                // assert type
                let v = <AnnotationDef as Into>::into(v);
                vec.push(v);
            }
            Ok(vec)
        }
    }

    deserializer.deserialize_seq(AnnotationDefVisitor)
}
```

Then you'd use `#[serde(deserialize_with = "deserialize_annotation_vec"]` on the field to use this implementation.

Sources:

- deserialize method: https://serde.rs/custom-date-format.html
- visitor: https://serde.rs/impl-deserialize.html
- Vec visitor impl (for body of `visit_seq`): https://github.com/serde-rs/serde/blob/master/serde/src/de/impls.rs#L625

**Comment by dtolnay** [comment]
I would implement it as:

```rust
use serde::{Deserialize, Deserializer};

#[derive(Deserialize)]
#[serde(remote = "Snippet")]
struct SnippetDef {
    #[serde(deserialize_with = "vec_annotation")]
    annotations: Vec<Annotation>,
}

fn vec_annotation<'de, D>(deserializer: D) -> Result<Vec<Annotation>, D::Error>
where
    D: Deserializer<'de>,
{
    #[derive(Deserialize)]
    struct Wrapper(#[serde(with = "AnnotationDef")] Annotation);

    let v = Vec::deserialize(deserializer)?;
    Ok(v.into_iter().map(|Wrapper(a)| a).collect())
}
```

**Comment by zbraniecki** [comment]
Thank you! That works great!

**Comment by dtolnay** [comment]
I believe those were stabilized already in 1.27.0. I don't see how it would apply to this issue though.

**Comment by RReverser** [comment]
Hmm, you're right, I misunderstood what it does. Allowing them on generic params doesn't mean getting attributes for these params from instantiation sites in a generic one.

**Comment by stevenroose** [comment]
Still no solution or workaround for the `Option<>` case??

I'm doing his because I have a lot of `Vec<u8>`s that are encoded as hex. So I use `deserialize_with = "deserialize_hex"` with a custom hex method. However, now I have `Option<Vec<Vec<u8>>>` and I lost it all. I made a new `deserialize_hex_array` which can decode `Vec<Vec<u8>>`, but with the option it stops working. I tried having `deserialize_hex_array`return the `Option<Vec<Vec<u8>>>`, but still it didn't seem to work. I still got "missing field".

**Comment by daboross** [comment]
@stevenroose I know this is another workaround, but still. For missing fields like that I've found using `#[serde(default)]` along with the "option" deserialize_with works well.

I have a ton of fields that are `#[serde(default, deserialize_with = "option_timestamp")]` for example.

**Comment by markazmierczak** [comment]
I invested some time yesterday to tackle this and maybe I am on something.
Potentially, this is my first contribution to Rust community... I am excited !

TL;DR

I have a proof-of-concept implementation here: [Gist](https://gist.github.com/markazmierczak/f504f40c286ecd9f3204972c950961f3) (do not be frightened by the amount of code). The procedural macro has not been altered to support `as`/`serialize_as`/`deserialize_as` attributes, but this is easy part, I think.

The idea I have may enable us to write:

* ```rust
  #[derive(Serialize, Deserialize)
  struct SomeTime {
      #[serde(as = "Option<chrono::DateTime<chrono::Utc>>")]
      stamp: Option<chrono::NaiveDateTime>,
  }
  ```

  The above example is useful when `NaiveDateTime` (UTC) is used internally and the conversion between ISO8061 is needed only at API side.

* ```rust
  #[derive(Serialize, Deserialize)
  struct SomeTime {
      #[serde(as = "Vec<Hex>")]
      bytes: Vec<Vec<u8>>,
  }
  ```

  Note how type nesting plays nicely here.

* Any container can be supported:
  ```rust
  #[derive(Serialize, Deserialize)
  struct SomeTime {
      #[serde(as = "HashMap<i32, Option<chrono::DateTime<chrono::Local>>>")]
      map: HashMap<i32, Option<chrono::NaiveDateTime>>,
  }
  ```

As you can see:

* the syntax is easy
* the intent is very clear.
* the composition is great. You can nest types however you want.

# Details

Often one wants to serialize a struct field using specific format.
Changing a type of this field is often undesirable, because it requires more code to pack new type and unpack original type from new type. And `serde`'s job is nothing but conversions. The sad story is `serde` doesn't help us in this specific conversion.

As many already found `serialize_with` and `deserialize_with` have shortcomings.
They do not compose well and we quickly come into explosion of conversion functions for each variant
`Option<T>`, `Vec<Option<T>>`, `HashMap<K, Option<V>>` and so on.

With `Deserialize` and `Serialize` traits, we **do not have the ability to encode how nested types should be serialized**.
What I propose is to add new traits with such an ability:

```rust
/// Deserialize `T` as it would be `Self` type.
pub trait DeserializeAs<'de, T>: Sized {
    fn deserialize_as<D>(deserializer: D) -> Result<T, D::Error>
    where
        D: Deserializer<'de>;

    // TODO: deserialize_as_into
}

/// Serialize `T` as it would be `Self` type.
pub trait SerializeAs<T> {
    fn serialize_as<S>(source: &T, serializer: S) -> Result<S::Ok, S::Error>
    where
        S: Serializer;
}
```

This can be implemented between any two types we want, similar to `From` trait.
In the Gist you can find some implementations, implementations for standard collections are also included.

The good part is any crate can implement this trait for:

* non-standard collections/containers
* usual conversions (e.g. `chrono::NaiveDateTime` -> `chrono::DateTime`, `Vec<u8>` -> `Hex`, `Vec<u8>` -> `Base64`)

By doing this we can share more code within community.

The traits are implemented for more specific type, not the other way around.
Otherwise one would not be able to implement conversion for its own type due to coherence rules:

```rust
impl SerializeAs<Hex> for String { .. } // would not compile
impl SerializeAs<String> for Hex { .. }
```

The above is compatible with the order of `From` trait.

This is a big change and I am not sure whether this is preferred way of implementing this feature.
If it is, I definitely want to take this forward. But I will need help, some guidance.

I thought about externalizing this change to `serde-as` crate, but I think this would
reduce the adoption.

## SameAs

There is one caveat, though.
If leaf type does not change, it does not compile.
The example with `HashMap` is an instance of this problem, where `i32` key does not change.

The problem arises, because we can't implement:

```rust
impl<T: Serialize> SerializeAs<T> for T { .. }
```

We can't implement it, because the compiler cannot prove the implementations are not overlapping.

That's why I introduced `SameAs` structure, as a workaround.

There are several solutions:

* resolve https://github.com/rust-lang/rfcs/issues/1834
* implement traits for for all types manually
* **the most viable** solution right now:
  * move `SameAs` to `private` module
  * detect equivalence of leaf types in procedural macro (is it even possible to parse generic types ?).
    I know this will easily break as we cannot assume `String` is equal to `std::string::String`, but the chance one would write different types is low.
* make `SameAs` truly public and enforce everyone to write `serde(as = HashMap<SameAs<i32>, ..>)`

**Comment by dtolnay** [comment]
This looks promising! I like how well it composes for nested types.

I don't necessarily want to start by landing new traits and derive attributes in serde immediately. It would be better if all of this could be provided in a separate crate for now and we can iterate on the design.

For the attributes maybe you can rely on `with` for now:

```rust
use ???::As;

...

#[serde(with = "As<Vec<Hex>>")]
```

where `As` is a generic type that has serialize/deserialize methods as expected by `with` defined in terms of SerializeAs/DeserializeAs.

If someone already has a deserialize_with function (such as [this](https://github.com/cuechan/proxer-rs/blob/4e2978da6e173b3b9c6d395d90633d2ee751815f/src/response/info.rs#L14) / [this](https://github.com/cuechan/proxer-rs/blob/4e2978da6e173b3b9c6d395d90633d2ee751815f/src/response/mod.rs#L38), drawing arbitrarily from GitHub search results), could you show what they would need to write to make it work within this approach?

**Comment by markazmierczak** [comment]
`As` struct is great idea, I've tested it and there is only one little quirk.
It will be written as `#[serde(with = "As::<Vec::<Hex>>")]`, because serde expects path in this position.
I am going to write a new crate, yes!

The referenced `stringly_array_spaces` function could be rewritten without the use of `Visitor`. So let me do it first:

```rust
pub fn stringly_array_spaces<'de, D>(deserializer: D) -> Result<Vec<String>, D::Error>
where
    D: Deserializer<'de>,
{
    let s: &str = Deserialize::deserialize(deserializer)?;
    Ok(s.split_whitespace().map(|x| x.to_owned()).collect())
}
```

Translating this directly as it is to `DeserializeAs` would give us:

```rust
struct SpaceSeparatedStrings;

impl<'de> DeserializeAs<'de, Vec<String>> for SpaceSeparatedStrings {
    fn deserialize_as<D>(deserializer: D) -> Result<Vec<String>, D::Error>
    where
        D: Deserializer<'de>,
    {
        let s: &str = Deserialize::deserialize(deserializer)?;
        Ok(s.split_whitespace().map(|x| x.to_owned()).collect())
    }
}
```

This only give us the ability to nest `Vec<String>` inside containers, e.g. `Option`.

I would like to generalize this to take any parsable type (a type that implements `FromStr`).

```rust
struct SpaceSeparated<T>(PhantomData<T>);

impl<'de, T> DeserializeAs<'de, Vec<T>> for SpaceSeparated<T>
where
    T: std::str::FromStr,
    T::Err: fmt::Display,
{
    fn deserialize_as<D>(deserializer: D) -> Result<Vec<T>, D::Error>
    where
        D: Deserializer<'de>,
    {
        let s: &str = Deserialize::deserialize(deserializer)?;
        s.split_whitespace()
            .map(|x| T::from_str(x).map_err(<D::Error as serde::de::Error>::custom))
            .collect()
    }
}
```

But what I'm really missing here is a kind of `FromStrAs` to be able to fully utilize this idea.
Because, all we got now is only ability for outer nesting (inside other containers), but there is no way to redefine inner type, e.g.:

```rust
struct MyStruct {
    #[serde(as = "SpaceSeparated<Hex>")]
    member: Vec<Vec<u8>>,
}
```

As the consequence `SpaceSeparated`, as it stands now, doesn't even need to be parametrized by `T`.

**Comment by cedric-h** [comment]
@markazmierczak please publish the crate!

**Comment by dbeckwith** [comment]
I've been working around this problem for a while by writing deserializer functions for every combination of container-foreign type that I use. The `As` idea looks very promising!

**Comment by zicklag** [comment]
If I understand that right, that sounds pretty neat!

Would it allow me to do this?

```rust
// I have `Schedule` type that implements `Display` and `FromStr`.

#[derive(Serialize, Deserialize)]
struct Data {
    #[serde(with = "As<HashMap<Schedule, String>>")]
    cron_jobs: HashMap<String, String>
}
```

I would love to use that in my project right now. :slightly_smiling_face:

**Comment by kjeremy** [comment]
Is this available anywhere?

**Comment by cedric-h** [comment]
Not to my knowledge @kjeremy

**Comment by jonasbb** [comment]
I tested the code a bit and so far it felt quite nice to use. I'm considering making this the basis for my crate [serde_with](https://github.com/jonasbb/serde_with), in which I already have different serde helpers. I'm tracking the progress at https://github.com/jonasbb/serde_with/issues/87

A couple of notes though. The type `SameAs<T>` is more complicated than needed. A `Same` without explicit generics also works, but it might be nice to be more explicit.

While a generic implementation like
```rust
impl<T: Serialize> SerializeAs<T> for T { .. }
```
is not possible, it is possible to implement it for every `T` manually, thus maybe avoiding the need for `SameAs<T>`/`Same` in common cases like
```rust
impl SerializeAs<i32> for i32 { .. }
```

**Comment by michaelvoronov** [comment]
Hi, what is the current status of this feature? It could simplify my code a lot :).

**Comment by armansito** [comment]
Hello! Any updates on this feature? It would definitely improve the ergonomics to have something like this.

**Comment by alexdevteam** [comment]
Any progress on this? I've tried to use the `serde_with` crate, but i can't find any documentation explaining how to use it with remote types. I have a `Vec` of a remote type and I just can't figure out how to serialize it without doing it manually.

**Comment by jonasbb** [comment]
@alexdevteam I'd love to hear in more details what you tried with `serde_with` and what didn't work. Please post it [here](https://github.com/jonasbb/serde_with/issues/new).

**Comment by ColmanHumphrey** [comment]
Recently I tried to use @dtolnay's excellent [string_or_struct](https://serde.rs/string-or-struct.html) function on a vector. 

Building on David's code, let's assume we have instead:
```rust
#[derive(Deserialize, Serialize)]
pub struct ManyBuilds {
    some_fields_with_overall_info: i32, // etc
    // ideally we'd have #[serde(deserialize_with = "string_or_struct") or other notation
    builds: Vec<Build>
}
// see https://serde.rs/string-or-struct.html for Build's definition and traits
```

I tried @dtolnay's [also-excellent wrap function](https://github.com/serde-rs/serde/issues/723#issuecomment-382501277) from this thread, although I got generic and lifetime errors that I don't yet have the skills to resolve (I'll try again!). I also couldn't find how to do this with `serde_with`, but let me know if there's a clear choice @jonasbb and I'd be happy to try and/or make an issue.

My current solution is a newtype:
```rust
#[derive(Deserialize, Serialize)]
pub struct ManyBuildsInput {
    some_fields_with_overall_info: i32, // etc
    builds: Vec<BuildWrap>
}

#[derive(Deserialize, Serialize)]
struct BuildWrap(#[serde(deserialize_with = "string_or_struct") Build)
```
Then implement From/TryFrom etc to and from the `ManyBuildsInput` struct to `ManyBuilds`. Extra work, but survivable!

(by the way - @dtolnay thanks for the crate, serde is amazing!)

**Comment by jonasbb** [comment]
@ColmanHumphrey I am happy for any feedback and if you need help using `serde_with` just open a new [discussion](https://github.com/jonasbb/serde_with/discussions). This should solve the problem:
```rust
    #[serde_as(as = "Vec<PickFirst<(_, DisplayFromStr)>>")]
    //                              ^ deserialize struct
    //                                 ^^^^^^^^^^^^^^ or FromStr
    builds: Vec<Build>
```

**Comment by ColmanHumphrey** [comment]
> @ColmanHumphrey I am happy for any feedback and if you need help using `serde_with` just open a new [discussion](https://github.com/jonasbb/serde_with/discussions). This should solve the problem:
> 
> ```rust
>     #[serde_as(as = "Vec<PickFirst<(_, DisplayFromStr)>>")]
>     //                              ^ deserialize struct
>     //                                 ^^^^^^^^^^^^^^ or FromStr
>     builds: Vec<Build>
> ```

Thanks @jonasbb - just created https://github.com/jonasbb/serde_with/discussions/333!

**Comment by vi** [comment]
Given `serde` already reached version 1, does this issue have a chance to be resolved by Serde itself (not by some external mean), preserving compatibility? 

Maybe inability to use containers should be documented explicitly, ideally with some special error message that points users to workarounds?

**Comment by DanielJoyce** [comment]
Just bumped into this as well, surprised it's still not handled.

**Comment by chanced** [comment]
Unless I'm missing something, this seems not only possible but fairly ergonomic? For example:

```rust
pub struct Jwk {
    #[serde(
        with = "b64::optional_seq_url_safe",
        skip_serializing_if = "Option::is_none",
        default
    )]
    pub x5c: Option<Vec<Vec<u8>>>,
}

pub(crate) mod b64 {
    pub(crate) mod optional_url_safe {
        #[cfg(not(feature = "std"))]
        use alloc::{string::String, vec::Vec};
        use base64::{engine::general_purpose::URL_SAFE_NO_PAD, Engine as _};
        use serde::{self, Deserialize, Deserializer, Serializer};

        pub fn serialize<S, T>(input: &Option<T>, serializer: S) -> Result<S::Ok, S::Error>
        where
            S: Serializer,
            T: AsRef<[u8]>,
        {
            if input.is_none() {
                return serializer.serialize_none();
            }
            serializer.serialize_str(&URL_SAFE_NO_PAD.encode(input.as_ref().unwrap().as_ref()))
        }

        pub fn deserialize<'de, D, T>(deserializer: D) -> Result<T, D::Error>
        where
            D: Deserializer<'de>,
            T: From<Option<Vec<u8>>>,
        {
            let val = Option::<String>::deserialize(deserializer)?;
            let val = val
                .map(|val| {
                    URL_SAFE_NO_PAD
                        .decode(val)
                        .map_err(serde::de::Error::custom)
                })
                .transpose()?;
            Ok(val.into())
        }
    }

    pub(crate) mod optional_seq_url_safe {
        #[cfg(not(feature = "std"))]
        use alloc::{string::String, vec::Vec};
        use base64::{engine::general_purpose::URL_SAFE_NO_PAD, Engine as _};
        use serde::{self, ser::SerializeSeq, Deserialize, Deserializer, Serializer};

        pub fn serialize<S, V, T>(input: &Option<T>, serializer: S) -> Result<S::Ok, S::Error>
        where
            S: Serializer,
            T: AsRef<[V]>,
            V: AsRef<[u8]>,
        {
            if input.is_none() {
                return serializer.serialize_none();
            }
            let input = input.as_ref().unwrap().as_ref();
            let mut ser = serializer.serialize_seq(Some(input.len()))?;
            for item in input {
                ser.serialize_element(&URL_SAFE_NO_PAD.encode(item.as_ref()))?;
            }
            ser.end()
        }

        pub fn deserialize<'de, D, T>(deserializer: D) -> Result<T, D::Error>
        where
            D: Deserializer<'de>,
            T: From<Option<Vec<Vec<u8>>>>,
        {
            let data = Option::<Vec<String>>::deserialize(deserializer)?;
            let data = data
                .map(|data| {
                    data.into_iter()
                        .map(|data| {
                            URL_SAFE_NO_PAD
                                .decode(data)
                                .map_err(serde::de::Error::custom)
                        })
                        .collect::<Result<Vec<_>, _>>()
                })
                .transpose()?;
            Ok(data.into())
        }
    }
}
```

**Comment by daboross** [comment]
The main problem with that is just that you have to write a custom module for every different modification. This issue is asking for a generic way to turn some serialization module that serializes/deserializes `X` into one that works on `Option<X>` or `Vec<X>`.

Doing it manually works fine, *if* you know what's needed. But if you're writing a carte for general use, what modules do you provide?

I've done this in the past for internal usage too, usualli having two modules, `do_thing` and `do_thing_optional`. It's just not general: this doesn't let you use `#[serde(with = "do_thing")]` for a `Vec<Thing>` or `HashMap,u8, Thing>`. You'd need to write two more modules to cover those two use cases.

It's annoying to write N modules for N different use cases, but I mean sure it works if you're the only user. It'd be much more out of hand if you were trying to write a library, though.

For my use case for this, https://github.com/daboross/serde-tuple-vec-map/, I've experimented with another option: just provide a wrapper type in addition to the main module. Have someone do `field: Option<serde_tuple_vec_map::Wrapper<Vec<(String, String)>>>` to work around the fact that they can't do `#[serde(with = "serde_tuple_vec_map")]` in that case. It still has a big downside, though, of requiring them to insert a wrapper type whose only purpose is interacting with `serde` into their data structure, and deal with it throughout their codebase.

**Comment by murar8** [comment]
Just released a small crate to help work around the issue: https://crates.io/crates/serde_nested_with
It will generate the workaround module for you.

**Comment by Pr0methean** [comment]
> This needs to support all the wrapper types too: Rc, Arc, Cell, RefCell, Mutex, RwLock.

Don't forget Box, Pin and Cow.
